src/tfr_navigation/src/navigator.cpp

```diff
@@ -0,0 +1,135 @@
+    goal_manager.initialize_goal();
+    //TODO send goal to navigation stack 
+
+    ros::Rate r(rate);  
+    r.sleep(); //this pause is for debugging only DELETE
+
+    bool success = true;
+    //test for completion
+    //TODO hook up to navigation stack here
+    while (true)
+    {
+        //Deal with preemption or error
+        if (server.isPreemptRequested() || !ros::ok()) 
+        {
+            //TODO hook up to navigation stack here
+            //cascade the preemption and wait
+            ROS_INFO("%s: preempted", action_name.c_str());
+            server.setPreempted();
+            success = false;
+            break;
+        }
+        //main case, update nav goal
+        else
+        {
+            if (goal_manager.location_code == tfr_msgs::NavigationGoal::TO_MINING)
+            {
+                //MAINTENENCE NOTE this is only safe to do if we are running on
+                //spin or spin once with no threaded callbacks
+                goal_manager.update_mining_goal(current_position.pose.pose);
+                //TODO send updated goal to navigation stack
+            }
+
+            //TODO hookup status query to navigation stack here
+
+            feedback.header.stamp = ros::Time::now();
+            feedback.header.frame_id = frame_id;
+            feedback.status = tfr_msgs::NavigationFeedback::OK;
+            //MAINTENENCE NOTE this is only safe to do if we are running on
+            //spin or spin once with no threaded callbacks
+            feedback.current = current_position.pose.pose;
+            feedback.goal = goal_manager.nav_goal.target_pose.pose;
+            server.publishFeedback(feedback);
+            ROS_INFO("servicing goal, %f", feedback.header.stamp.toSec());
+            r.sleep();
+            break; //debugging DELETE on integration
+
+
+        }
+    }
+
+    //TODO hook up to navigation stack here
+    if (success)
+    {
+        result.header.stamp = ros::Time::now();
+        result.header.frame_id = frame_id;
+        result.status = tfr_msgs::NavigationResult::OK;
+        //MAINTENENCE NOTE this is only safe to do if we are running on
+        //spin or spin once with no threaded callbacks
+        result.current = current_position.pose.pose;
+        result.goal = goal_manager.nav_goal.target_pose.pose;
+        server.setSucceeded(result);
+    }
+    else if (false)
+    {
+        //TODO hook error handling to navigation stack
+        result.status = tfr_msgs::NavigationResult::OK;
+        server.setAborted(result);
+    }
+    ROS_INFO("Navigation server finished");
+}
+
+
+/**
+ * Callback for updating the most recent position
+ * */
+void Navigator::update_position(const nav_msgs::OdometryConstPtr &msg)
+{
+    //get the pose without the covariance, not needed
+    //need to use shared pointer here for thread safety see:
+    //https://answers.ros.org/question/53234/processing-an-image-outside-the-callback-function/
+    current_position = *msg;
```

Comment:
This comment is a `TODO:`, right?  Why not convert it to a thread-safe type now so we don't forget?

Reply:
No I am dereferencing a boost shared pointer. This comment is to let you know that this is a fully thread safe operation, and no maitenence is required